### ISSUE TITLE
图解Python numpy基本操作 - 知乎

### ISSUE THREAD

**Issue by manipulative** [manipulative/collection#13]
本文很长，你忍一下。

Numpy是python的一个非常基础且通用的库，基本上常见的库pandas，opencv，pytorch，TensorFlow等都会用到。

Numpy的核心就是n维array，这篇文章将介绍一维，二维和多维array。

> Python是一种非常有趣且有益的语言，我认为只要找到合适的动机，任何人都可以熟练掌握它。但是要记住的是，如果你只想着凭借python去找一份工作的话，不是不行，但是很难。python这种语言更适合已经有一份工作的人，多学一个技能。  
> 可以从最简单也是最直观的数据分析学起来，并且试着从知乎出品的数据分析课开始。

python入门可以看这个详细回答：

Numpy与List的异同点
--------------

他俩非常相似，**同样都是容器，都能快速的取值的修改值，但是插入和删除会慢一点**。

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/388bc537-3c5a-4a9a-931a-c2aa686831d1.jpeg?raw=true)

Numpy的优点

*   更**紧凑**，特别是**多维数据**
*   当数据可以**向量化**的时候比list更快
*   通常是**同质化**的，数据相同时处理更快，比如都是浮点型或者整数型

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/10035d80-d6fc-4a37-90e0-bab81b758dff.jpeg?raw=true)

向量 Vector 或者一维向量 1D array
-------------------------

### **向量初始化**

通过list转化，自动变成np类型，**_shape_**为（3，）

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/1c1a836a-8fde-46a0-9086-db44c19d975e.png?raw=true)

！注意，如果list里面的值类型不相同，那么**_dtype_**就会返回”object“

如果暂时没有想要转化的list，可以全用0代替

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/8b8d97ce-e3f4-4afe-992e-081f0f69b17b.png?raw=true)

也可以复制一个已经存在的全0 向量

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/24f31e72-9c9e-409d-a0bc-aa6985ef4f1b.png?raw=true)

！注意，所有创建包含固定值_vector_的方法都有_**_like**_函数

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/af2b9404-1709-41ca-9a1e-56bce1472ae7.jpeg?raw=true)

还有经典的_arange_和_linspace_方法

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/4504d298-9328-4a8f-80eb-cddae6f16d50.jpeg?raw=true)

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/3070dd28-230c-4c48-8fe1-37e36f70d800.jpeg?raw=true)

！ **arange**方法对于数据类型敏感，比如**_arange_**(3)，**_dtype_** 为int，如果你需要float类型，可以_arange(3).astype(float)_

**生成随机array**

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/bfe65541-90e0-47c9-9719-00c7c58e14e4.jpeg?raw=true)

### **向量索引**

基础的向量索引操作，**只是展示部分数据，而不改变数据本身**

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/b172a07d-d9ba-4e8b-bedd-3c44be44a86a.jpeg?raw=true)

布尔操作

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/8e422485-2cdb-4b74-aa35-c0850f2fc9cc.jpeg?raw=true)

也可以用.**_where_** 和**_clip_**代替上面的方法

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/98be5eeb-b91b-4851-a694-f55031b5f2df.jpeg?raw=true)

### 向量操作

numpy的优势就是把vector当做数做整体运算，避免循环运算

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/748bbb21-5bff-4021-ac56-cb26ac8db188.png?raw=true)

*   \- \* /无所不能

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/0fcedcba-b04c-4136-8330-7759b08026a0.png?raw=true)

复杂的数学运算不在话下

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/455082ec-fb39-4d95-a72a-53b49d9ac43b.png?raw=true)

标量运算

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/0af91bf7-780a-4b74-89a2-1cfbb956fda0.png?raw=true)

三角函数

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/d9d962fe-f72c-4f0c-bef9-a691d0ad752d.jpeg?raw=true)

整体取整

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/01ea7090-45ee-4761-91c5-5af81892e568.png?raw=true)

numpy还可以做基础的统计操作，比如**_max，min， mean， sum_**等

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/7b242f68-876e-4ed4-8b84-ecd53a9d630f.jpeg?raw=true)

排序操作

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/65f9f59f-6829-40ca-8b9f-ddb8c34ec722.png?raw=true)

### 查找操作

numpy不像list有**_index_**函数，通常会用**_where_**等操作

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/6a81b041-a66e-437f-a1fb-921f0776b2a0.png?raw=true)

其中有三种方法：

1.  where，难懂且对于x处于array末端很不友好
2.  next，相对较快，但需要[numba](https://link.zhihu.com/?target=https%3A//numba.pydata.org/)
3.  searchsorted，针对于已排过序的array

二维array，也称matrix矩阵
------------------

初始化，注意「双括号」

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/00e135c5-ac4d-49e0-abe5-e66256c8cd24.jpeg?raw=true)

随机matrix，同一维类似

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/c1466f3f-85b5-44ad-95ea-0344117245b7.jpeg?raw=true)

索引操作，不改变matrix本身

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/48e9555a-54ac-4e24-8795-be940f163c17.jpeg?raw=true)

Axis 轴操作，在matrix中，axis = 0 代表列， axis = 1 代表行，默认axis = 0

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/39a0dea2-0a78-4ea6-99f0-9c9d6c124891.png?raw=true)

matrix算术 + - * / 和 ** 都可

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/ad3246f5-2010-4e36-8839-d2d3dbcc483d.png?raw=true)

也可以matrix与单个数，matrix与vector，vector与vector进行运算

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/4429c45a-cf8e-49ee-bf45-434aed86c566.jpeg?raw=true)

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/ae1253cb-8090-4f1d-8e32-ca76cb064da0.png?raw=true)

行向量 列向量

二维的转置如下，一维的也就是vector转置为自己本身

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/e57c1f71-5742-4a38-ad64-7aa916dd6a77.png?raw=true)

reshape改变形态

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/f4487732-b038-404f-b5d9-f171aa03134c.jpeg?raw=true)

自此，三种向量，一维array，二维列vector，二维行向量

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/173b86ee-7761-41c0-9193-3c451b04723e.jpeg?raw=true)

### 矩阵操作

合并matrix，hstack横向，vstack纵向，也可以理解为堆叠

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/d8d69240-8cb0-4f80-a450-f0bef347b76d.png?raw=true)

反向操作hsplit和vsplit

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/2e34a6fb-cca3-4b41-a011-da0ded8e9e88.jpeg?raw=true)

matrix的复制操作，tile整个复制，repeat可以理解为挨个复制

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/e7cf1e6d-4991-4e94-875d-e1b2bc0b4730.jpeg?raw=true)

delete删除操作

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/3b4d7398-bd7a-49da-9a12-bca868651356.jpeg?raw=true)

删除的同时也可以插入

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/5b343acc-2fdc-4923-bffd-b8b52bfeaee3.jpeg?raw=true)

append操作，只能在末尾操作

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/a0561be6-da15-49c2-907a-16e1711a993a.jpeg?raw=true)

如果只增加固定值，也可以用pad

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/14b707d8-b61d-4fb6-83be-0967919788d0.jpeg?raw=true)

网格化

c和python都很麻烦，跟别说再大点的数了

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/89a2d74f-f06e-45ba-a59d-ef071be080c7.png?raw=true)

采用类似MATLAB会更快点

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/f54b3ae7-27ad-4f0f-ab65-0ec383605a87.png?raw=true)

**当然numpy有更好的办法**

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/6c4d356f-911a-4ee0-ba35-a68e5d6ada12.png?raw=true)

### matrix统计

_sum,min,max,mean,median_等等

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/b33fbf73-d7c3-463e-b7fc-e4f0f27e8d82.png?raw=true)

_argmin_和_argmax_返回最小值和最大值的下标

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/253656ea-3511-4ed5-80f9-858fa9a4ca5b.jpeg?raw=true)

**_all_**和**_any_**也可以用

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/970e3636-3e93-49a0-940b-cc1ed806b1ec.jpeg?raw=true)

### matrix排序，注意axis

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/a5821fce-2d2b-45c2-8077-c4e67e3fb1c3.jpeg?raw=true)

3D array或者以上
------------

初始化，**_reshape_**或者硬来

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/e4777901-673d-4618-872d-2d06ff996de5.jpeg?raw=true)

可以考虑把数据抽象成一层层的数据

就像RGB值的图像一样

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/7b8961cc-5402-4d91-8a8d-3ec4b0e2b7ab.jpeg?raw=true)

跟1D和2D类似的操作，**_zeros_**， **_ones_**，**_rand_**等

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/05bfee06-0ef8-44ac-872c-9520b9771fcd.jpeg?raw=true)

**_vstack_**和**_hstack_**照样可以用，现在多了一个dstack，代表维度的堆叠

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/a0e795ff-88ec-4a3d-abfb-15c42867e520.jpeg?raw=true)

**_concatenate_**也有同样的效果

![](https://github.com/manipulative/_Images/blob/master/1-14-2023,%2021-20-03/fc253bf4-eb6b-404b-8967-0a745c9ba492.jpeg?raw=true)

总结：
---

本文总结了numpy对于1D，2D和多维的基本操作。
